lib/elixir/lib/macro.ex

```diff
@@ -164,6 +164,15 @@ defmodule Macro do
                          "the #{to_string call} operator can only take two arguments"
   end
 
+  # {:fn, _, _} is what we get when we pipe into an anonymous function without
+  # calling it, e.g., `:foo |> (fn x -> x end)`.
+  def pipe(expr, {:fn, _, _} = call_args, _integer) do
+    expr_str = to_string(expr)
+    raise ArgumentError, "cannot pipe #{expr_str} into an anonymous function" <>
+                         " without calling the function; use #{expr_str} |>" <>
```

Comment:
I would not use the `|>` operator in the error message because folks can call Macro.pipe with one of the others pipe looking operators available. I would say: "use (#{to_string call_args}).() or define the anonymous function as a regular private function".